### PR TITLE
add features for asynchronous downloads

### PR DESCRIPTION
**Pull request by jefleponot**
To solve asynschronous downloads that does work with phantomjs v2 (issue :  Synchronous XHR causing zero byte file downloads #1342 ), there is a proposal PR 

for @langholz and @istr , thanks to your specifications.

PS : operation synchronously javascript is simply an absurdity because it inhibits the language spirit.
PS : PS : Could it be possible that a great tools as CasperJS be written to use asynchronous methods
(http://phantomjs.org/api/webpage/method/evaluate-async.html)
We could use it in more domains like  "chrome extension tabs" : https://developer.chrome.com/extensions/tabs#method-executeScript

Best Regards 
jef

### CODE REVIEW

File: docs/modules/casper.rst [hunk 1]
```diff
@@ -688,10 +744,14 @@ Exits phantom with a logged error message and an optional exit status code::
...
-**Signature:** ``download(String url, String target[, String method, Object data])``
+**Signature:** ``download(String url, String target[, String method, Object data, Boolean async])``
 
 Saves a remote resource onto the filesystem. You can optionally set the HTTP method using the ``method`` argument, and pass request arguments through the ``data`` object (see `base64encode()`_)::
 
+.. note:: 
+
+since phantom 2.0.0 synchronous downloads are forbidden : By default async is true / For backward compatibilty async is false with phantom 1.x.x::
+
     casper.start('http://www.google.fr/', function() {
```
Comment (istr) [inline docs/modules/casper.rst]: I would never ever change the API behavior (i.e. default values) depending on the engine in use.

Mark this "obsolete / not recommended" in doc, throw a warning to the output when used and advise to use the newly created "waitForDownload" instead.

Better throw an error if used with an engine that does not support it than silently changing behavior here.

**Comment by istr** [comment]
@jefleponot Thanks a lot for this PR :+1: .

It seems that you [have a regression with form selects in your patch](https://travis-ci.org/n1k0/casperjs/builds/96003595). Some builds fail for receipt of JSON content.

Please:
- have a look into what causes the regression(s) and fix it
- lint the code (some builds fail due to linting)

I would never silently change the API behavior due to the engine version in use, though.
Could you re-work your PR [along my proposal](https://github.com/n1k0/casperjs/issues/1342#issuecomment-162535729) and
- just keep behavior for the sync functions
- mark them obsolete / not recommended in the doc and give a hint to use the new waitFor... function s instead?
- maybe throw an error within the non-recommended functions when used with an engine that does not support them?

I would really like to integrate this PR into 1.1 because it is a major improvement.

**Comment by jefleponot** [comment]
your are totaly right. You would never ever change the API behavior (i.e. default values) depending on the engine in use. The code will now throw an error.

I find my regression trouble with meld diff-tools.
I will do my best to make a PR soon.

Thanks for all your brillant analyses.

**Comment by jefleponot** [comment]
partial PR done. CU next time

**Comment by istr** [comment]
Nice! Thanks a lot. Restarted the failing job (known intermittent popup bug).

Will review, ask a second collaborator for opinion and then merge.

**Comment by jefleponot** [comment]
I have done a lot of improvements of CasperJS for my web service supervision job, that I newer commit. if you are interested in lots of PR, it will valid them.

**Comment by istr** [comment]
@jefleponot  Sounds good. Maybe it would be best to create separate feature branches to help make the different change-sets small.

Any improvement is welcome. Especially those that help to close the numerous open issues.

Thanks for your support.

File: docs/modules/casper.rst [hunk 2]
```diff
@@ -688,10 +749,16 @@ Exits phantom with a logged error message and an optional exit status code::
...
 -------------------------------------------------------------------------------
 
-**Signature:** ``download(String url, String target[, String method, Object data])``
+**Signature:** ``download(String url, String target[, String method, Object data, Boolean async])``
 
 Saves a remote resource onto the filesystem. You can optionally set the HTTP method using the ``method`` argument, and pass request arguments through the ``data`` object (see `base64encode()`_)::
 
+.. note:: 
+
+since phantom 2.0.0 synchronous downloads are forbidden : By default async is true / For backward compatibilty async is false with phantom 1.x.x::
```
Comment (mickaelandrieu) [inline docs/modules/casper.rst]: PhantomJS instead please :)

File: modules/casper.js [hunk 3]
```diff
@@ -254,11 +287,47 @@ Casper.prototype.back = function back() {
...
  * @param  String  data    The data to send, optional
+ * @param  Boolean async   Asynchroneous request? (default: false)
  * @return string          Base64 encoded result
  */
-Casper.prototype.base64encode = function base64encode(url, method, data) {
+Casper.prototype.base64encode = function base64encode(url, method, data, async) {
     "use strict";
-    return this.callUtils("getBase64", url, method, data);
+    async = ((typeof(async) !== "undefined")? async : false);
+	if ((phantom.version.major > 1) && (!async)) {
```
Comment (mickaelandrieu) [inline modules/casper.js]: I'd like to directly set async to true if phantomjs is > 2 instead of throw an error.

File: modules/casper.js [hunk 4]
```diff
@@ -254,11 +287,47 @@ Casper.prototype.back = function back() {
...
+	   throw new CasperError("async parameter must be 'true' with base64encode");
+    }        
+    var dl = {"url": url}, dls = this.ajax.filter( function searchDownload( ajax ){
```
Comment (mickaelandrieu) [inline modules/casper.js]:
```js
var dl = {},
    dls = ...
```

File: modules/casper.js [hunk 5]
```diff
@@ -254,11 +287,47 @@ Casper.prototype.back = function back() {
...
+		   return (ajax.url === url) && (typeof (ajax.filename) === "undefined");
+		});	
+	if (dls.length === 0) {
+		this.ajax.push(dl);
+	} else {
+		dl = dls[0];
+	}    
+    dl.content = this.callUtils("getBase64", url, method, data, async);
```
Comment (mickaelandrieu) [inline modules/casper.js]: hmm I don't like the use of async as variable name, this could become a reserved keyword later...

how about ``asynchronous`` instead of ``async`` ?

File: modules/casper.js [hunk 6]
```diff
@@ -592,23 +661,67 @@ Casper.prototype.die = function die(message, status) {
...
  */
-Casper.prototype.download = function download(url, targetPath, method, data) {
+Casper.prototype.download = function download(url, targetPath, method, data, async) {
     "use strict";
     this.checkStarted();
-    var cu = require('clientutils').create(utils.mergeObjects({}, this.options));
-    try {
-        fs.write(targetPath, cu.decode(this.base64encode(url, method, data)), 'wb');
+    async = ((typeof(async) !== "undefined")? async : false);
+	if ((phantom.version.major > 1) && (!async)) {
```
Comment (mickaelandrieu) [inline modules/casper.js]: same here, let's limit obvious errors, and help users when we can :)

File: samples/download.js [hunk 7]
```diff
@@ -7,8 +7,17 @@
 
 var casper = require("casper").create();
 
-casper.start("http://www.google.fr/", function() {
-    this.download("http://www.google.fr/images/srpr/logo3w.png", "logo.png");
+casper.on('remote.message',function(msg){
+ console.log('message : '+msg);
+// fs.write('logo4.png', msg, 'wb');
```
Comment (mickaelandrieu) [inline samples/download.js]: remove uncommented code please

File: tests/suites/casper/logging.js [hunk 8]
```diff
@@ -37,6 +37,7 @@ casper.test.begin('logging tests', 4, function(test) {
         test.assert(casper.result.log.some(function(e) {
             return e.message === 'info message' && e.level === 'info' && e.space === 'remote';
         }), 'ClientUtils.log() adds a log entry at a given level');
+console.log(this.result.log.length);
```
Comment (mickaelandrieu) [inline tests/suites/casper/logging.js]: why do you need to console.log here ?

**Comment by mickaelandrieu** [comment]
@jefleponot great contribution ! some comments to address then. But I'm :+1: to merge this feature :)

File: modules/casper.js (hunk 3 above)
Comment (istr) [inline modules/casper.js]: I strongly disagree, see my comments to this PR. Implicitly changing behavior due to different engine in use is evil imho. Better to explicitly tell the user that what was planned is obsolete / unsupported right away.

File: modules/casper.js (hunk 6 above)
Comment (istr) [inline modules/casper.js]: Same here: let's tell the users "don't do this, use the way (async) that will be supported down the road". :)

**Comment by istr** [comment]
@mickaelandrieu thanks for quick review. :) Maybe let's discuss the question:
- silently change API behavior based on engine in use

vs.

- warn the user she's trying to do something that won't be supported in the future

approach with @n1k0, @hexid, and @paazmaya...

**Comment by jefleponot** [comment]
Another issue for you @istr : casperJS has never used "window.callPhantom" method before this PR.
It could be better to normalize this remote operation.

"window.callPhantom" is similar to the "window.postMessage" function : postMessage(data, targetDomain)

So it could be great to use the same object attributes 

The special propertys of a message event are:
<pre>
data
    The first argument of postMessage
origin
    The source domain
source
    The reference to sending window if it is possible to 
respond by calling event.source.postMessage
</pre>

chrome extensions has yet normalize communications with the "message passing" : 
https://developer.chrome.com/extensions/messaging
Message Passing is really useful to communicate between web page and "background page" (one to one) And also between many pages. It avoid to fall victim to cross-site scripting.

In the modules/casper.js file a senMessage method could be add like the <a href="https://developer.chrome.com/extensions/tabs#method-sendMessage">chrome.tabs.sendMessage </a> : sendMessage(integer tabId, any message, object options, function responseCallback)

Thanks for yours advices

**Comment by mickaelandrieu** [comment]
@istr sounds OK to me :+1:

**Comment by eloitay** [comment]
@jefleponot I tried patching this change into the source code myself and notice that it still produce zero byte files. So I am wondering if the way the script is written need to be changed as well in order for the bug to be fixed.

**Comment by jefleponot** [comment]
Hi @eloitay 
Could you send me you code ? I could help you more

Or you can do a PR with an example.

File: docs/modules/casper.rst [hunk 9]
```diff
@@ -410,7 +437,7 @@ Also have a look at `forward()`_.
 ``base64encode()``
 -------------------------------------------------------------------------------
 
-**Signature:** ``base64encode(String url [, String method, Object data])``
+**Signature:** ``base64encode(String url [, String method, Object data, Boolean async])``
```
Comment (paazmaya) [inline docs/modules/casper.rst]: This would give the impression that the third parameter is required when the second is given.
Perhaps `(String url [, String method [, Object data, Boolean async ] ] )`

Comment (jefleponot) [inline docs/modules/casper.rst]: thanks

**Comment by BIGjuevos** [comment]
Do we want to keep this issue in the 1.1 release @istr @mickaelandrieu? I vote to move to a new milestone 1.2.

I feel like this would be good for a 1.2 considering our focus on squashing bugs for 1.1.  This is something I'd be happy to pick up in the future.  Feel free to assign to me.

**Comment by mickaelandrieu** [comment]
Humm ok for 1.2, how many issues are blocking the release of 1.1 according to the actual milestone?

**Comment by jefleponot** [comment]
test ok - PhantomJS 1.9.7 -> 2.1.1
test ok - SlimmerJS 0.8.4 -> 0.9.6

result : 6748 octets
logo

code :
````js
var casper = require("casper").create();

casper.start("http://www.google.fr/", function() {
    this.download("http://www.google.fr/images/srpr/logo3w.png", "logo.png",'GET',{},true);
});
casper.then(function(){
    this.waitForDownload("http://www.google.fr/images/srpr/logo3w.png",function(){
        console.log('download sucess');
    }, function(){
        console.log('download failed');
    });
});
casper.run();
````

It is fully functional , works in production since January 2016 with phantomJS 2.1.1 because old download method can't be used

**Comment by jefleponot** [comment]
@mickaelandrieu : update done

**Comment by mickaelandrieu** [comment]
@jefleponot sadly, your last merged pull request have been merged and so on you need to rebase this one :dancer: 

@casperjs/contributors so, are we agree to merge this one: my vote is :+1:

**Comment by istr** [comment]
I would prefer not to break the API in this way:
- do not introduce boolean parameters
- especially do not introduce boolean parameters that change the previous "default" behaviour
- do not make the default behaviour / value dependent on the engine version in use

See https://github.com/casperjs/casperjs/issues/1342#issuecomment-207848750 what I would like to see in this PR instead:
- mark the current (broken) API functions obsolete and add a "broken" comment to the docs
- add a warning to those functions that say: "this is broken and will not work; it is obsolete and will be removed"
- create a *new* set of functions that implements what this PR achieves

**Comment by jefleponot** [comment]
@istr,

Thanks for your explanation.

you are right, this PR don't suit for many reasons.
JavaScript programming is now moving towards a purely asynchronous programming. 

> 
Synchronous XMLHttpRequest on the main thread is deprecated because of its detrimental effects to the end user's experience. For more help, check https://xhr.spec.whatwg.org/.

With PhantomJS and SlimerJS, on the opposite, methods are mixed synchronous or asynchronous : 
- open(url, callback) - **_asynchronous_**
- evaluateAsync(function, [delayMillis, arg1, arg2, ...]) - **_asynchronous_**

-  ==> page.evaluate(function, arg1, arg2, ...) - **synchronous**
- ==> page.render (filename [, {format, quality}]) - **synchronous**
- ==> page.renderBase64(format) - **synchronous**


If you look for an equivalent approach like chrome.tabs, every methods are fully asynchronous : 
- executeScript(integer tabId, object details, function **callback**)
- captureVisibleTab(integer windowId, object options, function **callback**)
(https://developer.chrome.com/extensions/tabs)

CasperJS has same trouble as PhantomJS and SlimerJS : mix synchronous and asynchronous fct.
wait(), waitFor(), waitForAlert(), waitForPopup() ...  - **_asynchronous_**
withFrame(), withPopup() - **_asynchronous_**

download() - **synchronous**
evaluate() - **synchronous**

The use of Synchronous mode, although it eases writing limits treatments.

@istr, could we propose a new set of asynchronous functions what could lead CasperJS to good javascript practices  ?

**Comment by mickaelandrieu** [comment]
@jefleponot @istr I agree with Jef, and we need to keep the old api.

Could be better for now to create new functions instead of improve the existing one (and this may be a good *start* for 2.0 api :) )

**Comment by sairoko12** [comment]
Someone what happened with this request?

**Comment by kensoh** [comment]
Archiving PR using **ARCHIVE** label as CasperJS, like PhantomJS, is not actively maintained for now. These PRs are from 1-5 years ago but there just hasn't been maintainers to continue the work.
